client/controller/web-util.js

```diff
@@ -22,6 +24,20 @@ export function redirectLoggedIn( context, next ) {
 
 export function render( context ) {
 	ReactDom.render( context.layout, document.getElementById( 'wpcom' ) );
+
+	// Activate the next section after it's been successfully loaded.
+	// (`context.nextSectionToActivate` is set in client/sections-middleware.js)
+	// Section activation is triggered when the new section is rendered for the
+	// first time; this avoids glitches caused by the section's styles loading
+	// before the actual section's markup gets rendered..
+	if ( context.nextSectionToActivate ) {
+		context.store.dispatch( setSection( context.nextSectionToActivate ) );
+		context.store.dispatch( activateNextLayoutFocus() );
```

Comment:
After this patch, the `setSection` and `activateNextLayoutFocus` actions will be dispatched only after the `<Layout>` is rerendered with the _new section UI_, triggering a second rerender of the new UI immediately.

Before this patch, these dispatches triggered a rerender of the _old section UI_ before rendering the new one.

The `<Layout>` component uses the Redux state changed by these two actions: the `getSectionName` and `getSectionGroup` selectors and everything that depends on them, the `getCurrentLayoutFocus` selector...

I'm afraid that this patch fixes a set of visual glitches only at the cost of triggering other ones. The root cause -- rerendering the UI several times, with inconsistent Redux state and React UI (new section UI with old section state and vice versa) -- is not addressed at all.

Reply:
> The root cause -- rerendering the UI several times, with inconsistent Redux state and React UI (new section UI with old section state and vice versa) -- is not addressed at all.
That's correct — this PR does not try to solve the issue at its root. Doing so would be quite a big task, and require some planning ahead.

If I understand correctly, before these changes:
- UI still rendered at least twice (first with the old section's DOM, then with the new section's)
- the first render would cause glitches / bugs because of the mismatch between the new state and potentially old DOM

After these changes:
- UI still renders twice, but this time the changes caused by `setSection` should happen when the correct DOM has been rendered (for example, the focus could be moved correctly to a DOM element that is not going to be removed in the next render)
- there wouldn't be visual glitches

..to the net of new issues that haven't been found yet

Reply:
> the changes caused by `setSection` should happen when the correct DOM has been rendered

These changes happen only after the new DOM has been rendered, triggering a second rerender.

There is still a mismatch between state and DOM: the first render of the new DOM happens with the old state. For example, that can render a Signup screen with `.is-section-reader` CSS class. Only after the rerender the CSS class will be updated to `.is-section-signup`.

> there wouldn't be visual glitches

That's not guaranteed at all. The mismatch is still there, and can cause glitches, too.

Reply:
Makes sense, thank you for the detailed explanation